subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DependencyGraphBuilder.java

```diff
@@ -223,18 +231,29 @@ private void resolveEdges(final NodeState node,
         performSelectionSerially(dependencies, resolveState);
         computePreemptiveDownloadList(dependencies, dependenciesMissingMetadataLocally, componentIdentifierCache);
         downloadMetadataConcurrently(node, dependenciesMissingMetadataLocally);
-        attachToTargetRevisionsSerially(dependencies);
+        attachToTargetRevisionsSerially(dependencies, resolveState);
 
     }
 
-    private void attachToTargetRevisionsSerially(List<EdgeState> dependencies) {
+    private void attachToTargetRevisionsSerially(List<EdgeState> dependencies, ResolveState resolveState) {
         // the following only needs to be done serially to preserve ordering of dependencies in the graph: we have visited the edges
         // but we still didn't add the result to the queue. Doing it from resolve threads would result in non-reproducible graphs, where
         // edges could be added in different order. To avoid this, the addition of new edges is done serially.
         for (EdgeState dependency : dependencies) {
             if (dependency.getTargetComponent() != null) {
                 dependency.attachToTargetConfigurations();
             }
+            for (NodeState target : dependency.getTargetNodes()) {
+                List<? extends CapabilityDescriptor> capabilities = target.getMetadata().getCapabilitiesMetadata().getCapabilities();
```

Comment:
Rather than reaching exposing the `NodeState` list for each edge, it would be good to add `ComponentState.getCapabilities()`. A `ComponentState` represents the selected variant in the graph, which usually means a single `NodeState`/`ConfigurationMetadata`, but at times it will have multiple.